storage/namespace.go

```diff
@@ -728,6 +732,55 @@ func (n *dbNamespace) Flush(
 	return res
 }
 
+func (n *dbNamespace) Snapshot(blockStart, snapshotTime time.Time, flush persist.Flush) error {
+	// NB(rartoul): This value can be used for emitting metrics, but should not be used
+	// for business logic.
+	callStart := n.nowFn()
+
+	n.RLock()
+	if n.bs != bootstrapped {
+		n.RUnlock()
+		n.metrics.snapshot.ReportError(n.nowFn().Sub(callStart))
+		return errNamespaceNotBootstrapped
+	}
+	n.RUnlock()
+
+	if !n.nopts.SnapshotEnabled() {
+		n.metrics.snapshot.ReportSuccess(n.nowFn().Sub(callStart))
+		return nil
+	}
+
+	multiErr := xerrors.NewMultiError()
+	shards := n.GetOwnedShards()
+	for _, shard := range shards {
+		isSnapshotting, lastSuccessfulSnapshot := shard.SnapshotState()
+		if isSnapshotting {
+			// Should never happen because snapshots should never overlap
+			// each other (controlled by loop in flush manager)
+			n.log.
+				WithFields(xlog.NewField("shard", shard.ID())).
+				Errorf("tried to snapshot shard that is already snapshotting")
```

Comment:
We're accumulating a lot of these `// should never happen` (which is great) in the code base. I'm a little concerned we won't won't be able to tell the severity of such a message purely by looking at the log file. Could we put something in these messages which makes it apparent that it's a big deal to see the message, something like:

```
Errorf("[invariant violated] tried to snapshot shard that is already snapshotting")
```

Reply:
maybe also open an issue to track this change for other callsites

Reply:
done: #519